# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

## Step 1. Open the data file and have a look at the general information. 
#### Description of the data
- `children`: the number of children in the family

- `days_employed`: how long the customer has been working

- `dob_years`: the customer’s age

- `education`: the customer’s education level

- `education_id`: identifier for the customer’s education

- `family_status`: the customer’s marital status

- `family_status_id`: identifier for the customer’s marital status

- `gender`: the customer’s gender

- `income_type`: the customer’s income type

- `debt`: whether the customer has ever defaulted on a loan

- `total_income`: monthly income

- `purpose`: reason for taking out a loan

In [13]:
from __future__ import division
from IPython.display import display 
import pandas as pd

try:
    credit_scoring = pd.read_csv('credit_scoring_eng.csv')
except:
    credit_scoring = pd.read_csv('/datasets/credit_scoring_eng.csv')
credit_scoring.info()
print()
print()
credit_scoring_missing = credit_scoring.isna()
credit_scoring_missing_num = credit_scoring_missing.sum()
print('What are the columns with the missings:\n{}'.format(credit_scoring.loc[:, credit_scoring.isnull().any()].columns))
print()
print('Total number of missing values in each column:\n{}'.format(credit_scoring_missing_num))
print()
print('Number of rows in our DataFrame:\n{:}'.format(len(credit_scoring)))
print()
print("percentage of missing value in days_employed column:\n{:.2%}".format(credit_scoring_missing_num['days_employed']/len(credit_scoring))) 
print()
print("percentage of missing value in total_income column:\n{:.2%}".format(credit_scoring_missing_num['total_income']/len(credit_scoring))) 
print()
print(credit_scoring.describe())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


What are the columns with the missings:
Index(['days_employed', 'total_income'], dtype='object')

Total number of missing values in each column:
children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender

In [4]:
#A general view to 'days_employed' column
display(credit_scoring)
print(credit_scoring['days_employed'].head())
print()
print(credit_scoring['days_employed'].tail())
print()
print('Apparently we have many negative values in days_employed column, lets check:\n{:}'.format(credit_scoring[credit_scoring['days_employed'] < 0]['days_employed'].count()))
print()
print('Positive values?\n{:}'.format(credit_scoring[credit_scoring['days_employed'] >= 0]['days_employed'].count()))
print()
print('Lets see if we have meaningnfull values regarding dob_years column\n')
print('Calculating mean of days_employed per years\n{:}'.format(abs((credit_scoring['days_employed'].mean()))/365))
print()
print('Calculating mean of dob_years\n{:}'.format(credit_scoring['dob_years'].mean()))
print()
print('Lets see now for each negative and positive values')
print('Positive values mean per years\n{:}'.format(credit_scoring[credit_scoring['days_employed'] >= 0]['days_employed'].mean()/365))
print()
print('Negative values mean per years\n{:}'.format(abs(credit_scoring[credit_scoring['days_employed'] < 0]['days_employed'].mean())/365))


children  days_employed  dob_years            education  education_id  \
0             1   -8437.673028         42    bachelor's degree             0   
1             1   -4024.803754         36  secondary education             1   
2             0   -5623.422610         33  Secondary Education             1   
3             3   -4124.747207         32  secondary education             1   
4             0  340266.072047         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0         40620.102    purchase of the house  
1         17932.802             car purchase  
2         23341.752    purchase of the house  
3         42820.568  supplementary education  
4         25378.572        to have a wedding  
...             ...                      ...  
21520     35966.698     housing transactions  
21521     24959.969        purchase of a car  
21522     14347.610                 property  
21523     39054.888        buying my own car  
21524     13127.587             to buy a car  

[21525 rows x 12 columns]

0     -8437.673028
1     -4024.803754
2     -5623.422610
3     -4124.747207
4    340266.072047
Name: days_employed, dtype: float64

21520     -4529.316663
21521    343937.404131
21522     -2113.346888
21523     -3112.481705
21524     -1984.507589
Name: days_employed, dtype: float64

Apparently we have many negative values in days_employed column, lets check:
15906

Positive values?
3445

Lets see if we have meaningnfull values regarding dob_years column

Calculating mean of days_employed per years
172.73013057937914

Calculating mean of dob_years
43.29337979094077

Lets see now for each negative and positive values
Positive values mean per years
1000.011807989777

Negative values mean per years
6.446618991777744


In [5]:
print('Lets see if we have any duplicate rows??\n{:}'.format(credit_scoring.duplicated().sum()))

Lets see if we have any duplicate rows??
54


### Conclusion

- Missing values?  We have 2174  NA in 'days_employed' and 'total_income' columns (nearly 10% for each of the 2 col). 
- 'days_employed' column contains many negative value which is inaccurate.
- There is a huge gap between 'data_employed' "positive values" and 'dob_years' (the days employed are so much bigger than customer's ages).
- Also, the mean of data_employed positive values per year is illogical.
- 54 duplicate rows.

## Step 2. Data preprocessing

### Processing missing values

In [6]:
print('Identifying what family status we do have:\n{:}'.format(credit_scoring['family_status'].unique()))
print()
print('Identifying what family status ids we do have:\n{:}'.format(credit_scoring['family_status_id'].unique()))
print()
print('Well I think those two columns stand for the same thing!')

Identifying what family status we do have:
['married' 'civil partnership' 'widow / widower' 'divorced' 'unmarried']

Identifying what family status ids we do have:
[0 1 2 3 4]

Well I think those two columns stand for the same thing!


In [28]:
#now lets see what to do about our missing values!
#'days_employed' and 'total_income' are quantitative columns, so we can replace missing data with representative values!
#lets start with analysing 'total_income' since it only has missing values. Any relation with other columns???
#apparently we have 'income_type', lets see:
print(credit_scoring.groupby('income_type')['total_income'].sum())

income_type
business                       1.600214e+08
civil servant                  3.928580e+07
employee                       2.842091e+08
entrepreneur                   1.030690e+05
paternity / maternity leave    8.612661e+03
retiree                        8.512356e+07
student                        1.571226e+04
unemployed                     4.202872e+04
Name: total_income, dtype: float64


In [25]:
#we can see here that 'total_income' has some serious outliers. We can use the median to fill the missings.
median_income = credit_scoring['total_income'].median()
credit_scoring['total_income'] = credit_scoring['total_income'].fillna(value = median_income).reset_index(drop=True)
print('Total number of missing values in each column:\n{}'.format(credit_scoring.isna().sum()))
#ok, cool!

Total number of missing values in each column:
children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dtype: int64


In [54]:
#lets see what to do with 'days_employed' now!
#we already found some negative values in this column, lets check other columns.
quantitative_cols = ['children','dob_years','education_id','family_status_id','debt','total_income']
display_count_zero = [credit_scoring[credit_scoring[c] < 0][c].count() for c in quantitative_cols]
for i in range(len(quantitative_cols)):
    print(quantitative_cols[i]+' '+str(display_count_zero[i]))

children 0
dob_years 0
education_id 0
family_status_id 0
debt 0
total_income 0


In [39]:
#as we see over here, 'children' contains negative values too, which is impossible.
credit_scoring['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [44]:
#checking for 'dob_years'...
credit_scoring['dob_years'].value_counts()
#I can see a 110 '0' here, 
print(credit_scoring[credit_scoring['dob_years'] == 0]['dob_years'].count())

101


- Well, I'm not sure about na values, since both 'days_employed' and 'total_income' had the same nbre of na values..
- Both columns are quantitative data, we can replace NAs with one of the representative values such as median or mean. And the outliers in 'total_income' column are very significative, so teh best is choosing median() fct.
- Negative values in 'children' column.
- Zeros in 'dob_years' !!
- Negative values in 'days_employed' column, I'm thinking about converting this data... let's see!

I'm going to check all different columns.



In [45]:
#replacing negative values in 'children' column!
median_children = credit_scoring['children'].median()
credit_scoring.loc[credit_scoring['children'] < 0,'children'] = median_children
credit_scoring['children'].value_counts()
#okay!

0.0     14196
1.0      4818
2.0      2055
3.0       330
20.0       76
4.0        41
5.0         9
Name: children, dtype: int64

In [47]:
#replacing null values in 'dob_years' column!
median_dobyears = credit_scoring['dob_years'].median()
credit_scoring.loc[credit_scoring['dob_years'] == 0,'dob_years'] = median_dobyears
credit_scoring['dob_years'].value_counts()
#okay!

42.0    698
35.0    617
40.0    609
41.0    607
34.0    603
38.0    598
33.0    581
39.0    573
31.0    560
36.0    555
44.0    547
29.0    545
30.0    540
48.0    538
37.0    537
50.0    514
43.0    513
32.0    510
49.0    508
28.0    503
45.0    497
27.0    493
56.0    487
52.0    484
47.0    480
54.0    479
46.0    475
58.0    461
57.0    460
53.0    459
51.0    448
59.0    444
55.0    443
26.0    408
60.0    377
25.0    357
61.0    355
62.0    352
63.0    269
64.0    265
24.0    264
23.0    254
65.0    194
66.0    183
22.0    183
67.0    167
21.0    111
68.0     99
69.0     85
70.0     65
71.0     58
20.0     51
72.0     33
19.0     14
73.0      8
74.0      6
75.0      1
Name: dob_years, dtype: int64

In [55]:
#looking now for qualitative data, if there is any processing needed, lets see!
credit_scoring['education'].value_counts()
#hmmm, some format standardization must be done here.

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

In [66]:
credit_scoring['education'] = credit_scoring['education'].str.lower()

In [67]:
credit_scoring['education'].value_counts()

secondary education    15233
bachelor's degree       5260
some college             744
primary education        282
graduate degree            6
Name: education, dtype: int64

In [56]:
#family_status
credit_scoring['family_status'].value_counts()

married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: family_status, dtype: int64

In [57]:
credit_scoring['family_status_id'].value_counts()
#btw, this is a duplicated column ('family_status'/'family_status_id')

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

In [60]:
credit_scoring['gender'].value_counts()
#XNA?? ok ok! I should replace this with... ??? 

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [68]:
credit_scoring.loc[credit_scoring['gender'] == 'XNA','gender'] = 'F' #Voila!

In [63]:
credit_scoring['gender'].value_counts()

F    14237
M     7288
Name: gender, dtype: int64

In [64]:
credit_scoring['income_type'].value_counts()
#nothing to fix

employee                       11119
business                        5085
retiree                         3856
civil servant                   1459
entrepreneur                       2
unemployed                         2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64

In [65]:
credit_scoring['purpose'].value_counts()
#sounds good!

wedding ceremony                            797
having a wedding                            777
to have a wedding                           774
real estate transactions                    676
buy commercial real estate                  664
housing transactions                        653
buying property for renting out             653
transactions with commercial real estate    651
purchase of the house                       647
housing                                     647
purchase of the house for my family         641
construction of own property                635
property                                    634
transactions with my real estate            630
building a real estate                      626
buy real estate                             624
purchase of my own house                    620
building a property                         620
housing renovation                          612
buy residential real estate                 607
buying my own car                       

### Conclusion

- We've done now with NAs, insignificant values, for qualitative and quantitatives.
- I'm thinking about droping the 'family_status_id' column.
- Let's see what can I do else with this data!

**PS: I forgot about 'days_employed'!!**
I don't think there are much choices, I'm thinking about transforming it into absolute values, let's see.

In [76]:
credit_scoring.loc[credit_scoring['days_employed'] < 0,'days_employed'] = abs(credit_scoring['days_employed'])

In [78]:
print(credit_scoring.loc[credit_scoring['days_employed'] < 0,'days_employed'].sum())

0.0


## Data type replacement

In [79]:
credit_scoring.dtypes

children            float64
days_employed       float64
dob_years           float64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income        float64
purpose              object
dtype: object

### Conclusion

Nothing to fix in data types I think! Hmmm, I can for example convert some float64 to int64? 
- For 'children' column for example, no meaning to have a float here.

In [91]:
credit_scoring['children'] = credit_scoring['children'].astype('int64', errors='ignore')

In [92]:
credit_scoring['children'].dtypes

dtype('int64')

## Processing duplicates

Checking if there are any duplicates?

In [94]:
credit_scoring.duplicated().sum()

72

In [97]:
#droping duplicates
credit_scoring = credit_scoring.drop_duplicates().reset_index(drop=True)
credit_scoring.duplicated().sum()

0

### Conclusion

Let's try to lower or upper case all qualitative data, and reverify about duplicates.

In [98]:
credit_scoring['family_status'] = credit_scoring['family_status'].str.lower()
credit_scoring['gender'] = credit_scoring['gender'].str.upper()
credit_scoring['income_type'] = credit_scoring['income_type'].str.lower()
credit_scoring['purpose'] = credit_scoring['purpose'].str.lower()

In [99]:
credit_scoring.duplicated().sum()

0

OK! cool, next step!

## Lemmatization

'purpose'.........???

In [100]:
credit_scoring['purpose'].value_counts()

wedding ceremony                            791
having a wedding                            767
to have a wedding                           765
real estate transactions                    675
buy commercial real estate                  661
housing transactions                        652
buying property for renting out             651
transactions with commercial real estate    650
purchase of the house                       646
housing                                     646
purchase of the house for my family         638
construction of own property                635
property                                    633
transactions with my real estate            627
building a real estate                      624
buy real estate                             621
purchase of my own house                    620
building a property                         619
housing renovation                          607
buy residential real estate                 606
buying my own car                       

Just from the first sight, there are rows meaning the same thing.

In [136]:
import nltk
from nltk.stem import WordNetLemmatizer
from collections import Counter
#import stopwords with nltk
nltk.download('stopwords')
from nltk.corpus import stopwords
stop = stopwords.words('english')
wordnet_lemma = WordNetLemmatizer()
lemmas = []
for text in credit_scoring['purpose']:
    words = nltk.word_tokenize(text)
    lemmas = ' '.join([wordnet_lemma.lemmatize(w, pos = 'n') for w in words if w not in (stop)])
              
print(lemmas)
#print(Counter(lemmas))

[nltk_data] Downloading package stopwords to /home/jovyan/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


buy car


## Categorizing Data

In [ ]:
#droping 'family_status_id', but first drop duplicates
credit_scoring.drop_duplicates([''])

### Conclusion

## Step 3. Answer these questions

- Is there a relation between having kids and repaying a loan on time?

### Conclusion

- Is there a relation between marital status and repaying a loan on time?

### Conclusion

- Is there a relation between income level and repaying a loan on time?

### Conclusion

- How do different loan purposes affect on-time repayment of the loan?

### Conclusion

### Step 4. General conclusion

### Project Readiness Checklist

Put 'x' in the completed points. Then press Shift + Enter.

- [x]  file open;
- [x]  file examined;
- [x]  missing values defined;
- [x]  missing values are filled;
- [ ]  an explanation of which missing value types were detected;
- [ ]  explanation for the possible causes of missing values;
- [ ]  an explanation of how the blanks are filled;
- [ ]  replaced the real data type with an integer;
- [ ]  an explanation of which method is used to change the data type and why;
- [x]  duplicates deleted;
- [ ]  an explanation of which method is used to find and remove duplicates;
- [ ]  description of the possible reasons for the appearance of duplicates in the data;
- [ ]  data is categorized;
- [ ]  an explanation of the principle of data categorization;
- [ ]  an answer to the question "Is there a relation between having kids and repaying a loan on time?";
- [ ]  an answer to the question " Is there a relation between marital status and repaying a loan on time?";
- [ ]   an answer to the question " Is there a relation between income level and repaying a loan on time?";
- [ ]  an answer to the question " How do different loan purposes affect on-time repayment of the loan?"
- [ ]  conclusions are present on each stage;
- [ ]  a general conclusion is made.